engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/number/smallInteger/BitShiftNode.java

```diff
@@ -0,0 +1,122 @@
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.error.TypeError;
+import org.enso.interpreter.runtime.number.EnsoBigInteger;
+
+@ImportStatic(BigIntegerOps.class)
+@BuiltinMethod(type = "Small_Integer", name = "bit_shift", description = "Bitwise shift.")
+public abstract class BitShiftNode extends Node {
+  private @Child ToEnsoNumberNode toEnsoNumberNode = ToEnsoNumberNode.build();
+  private final ConditionProfile canShiftLeftInLongProfile =
+      ConditionProfile.createCountingProfile();
+  private final ConditionProfile positiveFitsInInt = ConditionProfile.createCountingProfile();
+  private final ConditionProfile negativeFitsInInt = ConditionProfile.createCountingProfile();
+  private final ConditionProfile rightShiftPositiveThis = ConditionProfile.createCountingProfile();
+  private final ConditionProfile bigIntNonNegative = ConditionProfile.createCountingProfile();
+  private final ConditionProfile rightShiftExceedsLongWidth =
+      ConditionProfile.createCountingProfile();
+
+  abstract Object execute(Object _this, Object that);
+
+  static BitShiftNode build() {
+    return BitShiftNodeGen.create();
+  }
+
+  @Specialization(guards = {"that >= 0", "canShiftLeftInLong(_this, that)"})
+  long doLongShiftLeft(long _this, long that) {
+    return _this << that;
+  }
+
+  @Specialization(guards = "that >= 0", replaces = "doLongShiftLeft")
+  Object doLongShiftLeftExplicit(
+      long _this, long that, @CachedContext(Language.class) ContextReference<Context> ctxRef) {
+    if (canShiftLeftInLongProfile.profile(canShiftLeftInLong(_this, that))) {
+      return doLongShiftLeft(_this, that);
+    } else if (positiveFitsInInt.profile(BigIntegerOps.fitsInInt(that))) {
+      return toEnsoNumberNode.execute(BigIntegerOps.bitShiftLeft(_this, (int) that));
+    } else {
+      throw new PanicException(
+          ctxRef.get().getBuiltins().error().getShiftAmountTooLargeError(), this);
+    }
+  }
+
+  @Specialization(guards = {"that < 0", "fitsInInt(that)"})
+  long doLongShiftRight(long _this, long that) {
+    if (rightShiftExceedsLongWidth.profile(-that >= 64)) {
+      if (rightShiftPositiveThis.profile(_this >= 0)) {
+        return 0L;
+      } else {
+        return -1L;
+      }
+    } else {
+      return _this >> -that;
+    }
+  }
+
+  @Specialization(guards = "that < 0", replaces = "doLongShiftRight")
+  long doLongShiftRightExplicit(long _this, long that) {
+    if (negativeFitsInInt.profile(BigIntegerOps.fitsInInt(that))) {
+      return doLongShiftRight(_this, that);
+    } else if (rightShiftPositiveThis.profile(_this >= 0)) {
+      return 0L;
+    } else {
+      return -1L;
+    }
+  }
+
+  @Specialization
+  Object doBigInteger(
+      long _this,
+      EnsoBigInteger that,
+      @CachedContext(Language.class) ContextReference<Context> ctxRef) {
+    if (bigIntNonNegative.profile(BigIntegerOps.nonNegative(that.getValue()))) {
```

Comment:
Don't profile this condition, instead reorder the branches such that this here is in `else`. There's no point predictively compiling a branch that refuses to compile as first statement.